### DISCUSSION TITLE
TCP Server is running on an Arduino - Clients are from a PC (using WatsonTCP)

### DISCUSSION THREAD

**Discussion by jimmie11** [jchristn/WatsonTcp#137] [Q&A]
Thank you very for making this great library.

I have a TCP server running on an Arduino. The Arduino TCP server will only be connected to Watson TCP clients running on a PC.

Can you help me with the code for the additional header on the Arduino side? I know it is explained in the link below but I do not know how to implement it on the Arduino side.

https://github.com/jchristn/WatsonTcp/blob/master/FRAMING.md#integration-with-non-watsontcp-endpoints

Thank you.

**Comment by jchristn** [comment]
Hi @jimmie11 thanks for your note.  What language are you using on the Arduino?  I assume Java.  While I can't produce the code for you, here's what I would do.

Receive:
1. Read one byte and append to the end of a buffer
2. If the end of the buffer equals ```\r\n\r\n``` you've reached the end of the message header, otherwise, go back to step 1
3. JSON-deserialize the buffer and convert the ```len``` property's value to an integer
4. Read ```len``` bytes from the socket

After step 4, you now have the entire message.

Send:
1. Create a ```HashMap``` (or similar dictionary) and set ```s``` to ```Normal```, and, ```len``` to the integer value representing the amount of data you wish to send
2. JSON serialize the ```HashMap``` to a string and send it to the socket (do not use JSON pretty which has newlines, tabs, etc)
3. Send ```\r\n\r\n``` to the socket
4. Send your data

Let me know if this helps.

**Reply by jimmie11** [reply]
Thank you very much Joel.  I really appreciate that you took the time to answer my question.

I am programming my Arduino in C (I do not know Java).  What I have done (but it is not working) is place the [WatsonTCP Start] bytes before my byte array (which is 266 bytes long).  Then, after my array, I placed [WatsonTCP End]  bytes as shown below.

[WatsonTCP Start]   [mybyteArray]   [WatsonTCP End] 

My code is below (Server side).  I would like to send this array to a WatsonTCP client and would appreciate if you can tell me why it is not working as the client is not receiving the array?  

Thanks in advance.

//---------------------------------------------------------------------------
```
#define MAX_TCP_LENGTH 285

  int msgS = 15;

  //Watson Start
  output[0] = 108;
  output[1] = 101;
  output[2] = 110;
  output[3] = 58;
  //------------------------
  output[4] = 01;
  output[5] = 10;
  //------------------------
  output[6] = 44;
  output[7] = 115;
  output[8] = 58;
  output[9] = 78;
  output[10] = 111;
  output[11] = 114;
  output[12] = 109;
  output[13] = 97;
  output[14] = 108;
//---------------------------------------------------------------------------
myArray(output[0 + msgS]    to  output[265 + msgS])
//---------------------------------------------------------------------------
//Watson End
  output[266 + msgS] = 13;
  output[267 + msgS] = 10;
  output[268 + msgS] = 13;
  output[269 + msgS] = 10;
```

**Reply by jchristn** [reply]
Hi @jimmie11 nothing will come after your byte array.  All of the Watson header information (including ```\r\n\r\n```) is before your byte array.

For instance:
```
{"s":"Normal","len":13}\r\n\r\nHello, world!
```
Would be a byte array containing: 
```
{"s":"Normal","len":13}
7b 22 73 22 3a 22 4e 6f 72 6d 61 6c 22 2c 22 6c 65 6e 22 3a 31 33 7d
```
then
```
\r\n\r\n
13 10 13 10
```
then
```
Hello, world!
48 65 6c 6c 6f 2c 20 77 6f 72 6c 64 21
```
So in total, you would send:
```
{"s":"Normal","len":13}\r\n\r\nHello, world!
7b 22 73 22 3a 22 4e 6f 72 6d 61 6c 22 2c 22 6c 65 6e 22 3a 31 33 7d 13 10 13 10 48 65 6c 6c 6f 2c 20 77 6f 72 6c 64 21
```
Let me know if this helps.  Cheers, Joel

**Reply by jimmie11** [reply]
T H A N K      Y O U  !

I appreciate it.  It works now after following your very clear directions.

**Reply by jimmie11** [reply]
Thanks again. Here is how the array should look like:

  //Watson Start
  output[0] = 123;
  output[1] = 34;
  output[2] = 115;
  output[3] = 34;
  output[4] = 58;
  output[5] = 34;
  output[6] = 78;     
  output[7] = 111;
  output[8] = 114;
  output[9] = 109;
  output[10] = 97;
  output[11] = 108;
  output[12] = 34;
  output[13] = 44;
  output[14] = 34;
  output[15] = 108;
  output[16] = 101;
  output[17] = 110;
  output[18] = 34;
  output[19] = 58;
  output[20] = 50;  //2     (message is 266 bytes)
  output[21] = 54 ; //6
  output[22] = 54 ; //6
  output[23] = 125; 
  output[24] = 13;
  output[25] = 10;
  output[26] = 13;
  output[27] = 10;

**Reply by jchristn** [reply]
Glad it worked!!!
